packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -212,19 +212,33 @@ class _TextEditingValueAccumulator {
   }
 }
 
-/// A [TextInputFormatter] that prevents the insertion of characters
-/// matching (or not matching) a particular pattern.
+/// A [TextInputFormatter] that prevents the insertion of characters matching
+/// (or not matching) a particular pattern, by replacing the characters with the
+/// given [replacementString].
 ///
 /// Instances of filtered characters found in the new [TextEditingValue]s
-/// will be replaced with the [replacementString] which defaults to the empty
+/// will be replaced by the [replacementString] which defaults to the empty
 /// string.
 ///
-/// Since this formatter only removes characters from the text, it attempts to
-/// preserve the existing [TextEditingValue.selection] to values it would now
-/// fall at with the removed characters.
+/// A [FilteringTextInputFormatter] also tries to preserve the existing
+/// [TextEditingValue.selection] and the [TextEditingValue.composing] region,
+/// adjusting them accordingly if text within either of these ranges is replaced.
```

Comment:
Nit: Is it also possible that that text before these ranges is replaced and they need to be updated?  Not sure but just being pedantic.